### PR TITLE
Autogenerate JSON Schema

### PR DESCRIPTION
**Pull request by alexiswl**
From #27 

## Summary

1. Run [typescript-json-schema](https://www.npmjs.com/package/typescript-json-schema) over the Workflow schema.  
2. Refine the schema (see [.github/scripts/curate_autogenerated_schema.py](https://github.com/alexiswl/cwl-ts-auto/blob/enhancement/generate-json-schema/.github/scripts/curate_autogenerated_schema.py))

## Schema refining process steps

1. Remove loadingOptions and extensionFields properties

2. Rename `class_`, `type_` etc to `class`, `type`

3. Add `$import` and `$include` options and update all requirements such that an item in a requirement can use $include or $import if that property accepts a list.

4. Fix inputs / outputs / steps, so not just an array but can be an object where the key is the `id` attribute.

5. Fix Requirements / Hints, so not just an array but can be an object where the key is the `class` attribute.

6. Fix descriptions by removing *Auto-generated...* from the start

7. Generate the CWLFile definition, which can be one of Workflow, CommandLineTool, or ExpressionTool and also fits the CWLMetadata

8. Fix issues with additionalProperties and nested definitions as shown here https://stoic-agnesi-d0ac4a.netlify.app/1 Needed for the CWLFIle and CWLGraphFile attributes

9. Generate the CWLGraph definition, which is an array of CWLFile objects with CWLMetadata

10. Write out the schema as JSON

11. To convert the schema to yaml, use the yq binary to generate the schema as yaml. ruamel.yaml doing some weird things on values under patternProperties regexes

### CODE REVIEW

File: .github/scripts/validate_schema.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+from tempfile import TemporaryDirectory
+from subprocess import run
+
+# Globals
+SCHEMA_PATH = Path('cwl_schema.json')
+SCHEMA_GRAPH_PATH = Path('cwl_schema.packed.json')
+CWL12_REPO_PARENT_TMPDIR_OBJ = TemporaryDirectory()
+CWL12_REPO_DIR = Path(CWL12_REPO_PARENT_TMPDIR_OBJ.name) / "cwl-v1.2"
+
+IGNORED_FILES = [
```
Comment (mr-c) [inline .github/scripts/validate_schema.py:22]: Thanks for checking against the conformance tests! I feel like there are more files to ignore.

Look for the `json_schema_invalid` tag https://github.com/search?q=repo%3Acommon-workflow-language%2Fcwl-v1.2%20json_schema_invalid&type=code

**Comment by alexiswl** [comment]
Moved back to graph and standard all-in-one schema, now auto-generates a yaml file as well, both files under `json_schema`

File: .github/scripts/curate_autogenerated_schema.py [hunk 2]
```diff
@@ -0,0 +1,1999 @@
...
+            ]
+        }
+
+    return schema_dict
+
+
+def fix_secondary_file_schema(schema_dict: Dict) -> Dict:
+    """
+    SecondaryFileSchema should allow for a single string entry to represent the secondary file pattern
+    Although not recommended, it is allowed in the CWL spec
```
Comment (mr-c) [inline .github/scripts/curate_autogenerated_schema.py:310]: Why is it not recommended?

File: .github/scripts/curate_autogenerated_schema.py [hunk 3]
```diff
@@ -0,0 +1,1999 @@
...
+            }
+        }
+    )
+
+    return schema_dict
+
+
+def fix_inline_javascript_requirement(schema_dict: Dict) -> Dict:
+    """
+    Fix the InlineJavascriptRequirement.expressionLib array to allow for $include
```
Comment (mr-c) [inline .github/scripts/curate_autogenerated_schema.py]: Shouldn't this be more universal, given all the places a `$include` can occur?

File: .github/scripts/curate_autogenerated_schema.py [hunk 4]
```diff
@@ -0,0 +1,1999 @@
...
+                        "type": "object"
+                    },
+                    # dct terms should also be permitted
+                    # see https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/metadata.cwl
+                    "^dct:.*$": {
+                        "type": "object"
+                    },
+                    "^http://purl.org/dc/terms/.*$": {
+                        "type": "object"
+                    }
```
Comment (mr-c) [inline .github/scripts/curate_autogenerated_schema.py]: it is valid CWL to have additional prefixed/namespaced  elements at any level, of any namespace; not just schema.org, DCT, etc..

Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py]: Okay, so could we use the regex `"^\w+:.*$"` for shorthand notation + `"^\w+:\/\/.*` for urls?

File: .github/scripts/curate_autogenerated_schema.py (hunk 3 above)
Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py]: What are all the places an include can occur? 

Can an $include occur in any requirement?

File: .github/scripts/curate_autogenerated_schema.py (hunk 2 above)
Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py:310]: It makes the requirement attribute ambiguous (and dependent on the implementation), success of the tool then depends on the workflow engine running the code -  See https://github.com/common-workflow-language/cwltool/issues/1322

File: .github/scripts/validate_schema.py (hunk 1 above)
Comment (alexiswl) [reply .github/scripts/validate_schema.py:22]: Are these 'json_schema_invalid' related to the inputs of the workflow or the workflow itself?

File: .github/scripts/curate_autogenerated_schema.py (hunk 4 above)
Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py]: I'm reluctant to just put in a `"^.*$"` for pattern properties here, it means spelling mistakes or syntax errors aren't necessarily picked up by the schema. `CWLVersion`, `cwlVersion`, `cwl_version` would all be valid entries

Comment (mr-c) [reply .github/scripts/curate_autogenerated_schema.py]: I agree that accepting any string anywhere is bad for validation. This is where having a full CWL language server like Benten can go beyond what JSON Schema can do. If you can allow keys that have a colon (`:`) in them, then that should suffice.

File: .github/scripts/curate_autogenerated_schema.py (hunk 3 above)
Comment (mr-c) [reply .github/scripts/curate_autogenerated_schema.py]: > Can an $include occur in any requirement?

Yes, and every other place: `$include` can be a key name at any level in a CWL document

https://www.commonwl.org/v1.2/Workflow.html#Document_preprocessing
https://www.commonwl.org/v1.2/SchemaSalad.html#Include

Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py]: Does this also apply for '$import'?

Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py]: If so, I'll just add them to the patternProperties of each of the definitions

Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py]: I've added $include and $import availabilities to all requirements (and hints), including top level of requirements 

So https://github.com/common-workflow-language/cwl-v1.2/blob/b60a42e3cc417c5b75b88fd7c6681abcc7ff5b89/tests/schemadef-wf.cwl#L7 would pass

as would https://github.com/common-workflow-language/cwl-v1.2/blob/b60a42e3cc417c5b75b88fd7c6681abcc7ff5b89/tests/schemadef_types_with_import-tool.cwl#L9

Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py]: Please see 19705a27178ef70534919ba724c65efdc1065d2c /  b31d8baea22b57e44c86dd8521eb03eba76e6ee6

File: .github/scripts/curate_autogenerated_schema.py (hunk 4 above)
Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py]: Resolved in 554c0f20262261bf72cff2278ffe319c7b67a91f / 986742842aea97c1b21a30caafda62f19bcbe36a

File: .github/scripts/curate_autogenerated_schema.py (hunk 2 above)
Comment (alexiswl) [reply .github/scripts/curate_autogenerated_schema.py:310]: Is this okay to keep in or should I remove the commentary?

File: .github/scripts/validate_schema.py (hunk 1 above)
Comment (alexiswl) [reply .github/scripts/validate_schema.py:22]: I think this one can be resolved

**Comment by alexiswl** [comment]
@mr-c please re-review when you have the chance, would be great to get this into the JSON schema store. Even if we just have this file generated into the cwl-ts-auto repository, we can generate a PR into the [JSON Schema store repository] (https://github.com/SchemaStore/schemastore) (like this one - https://github.com/SchemaStore/schemastore/pull/3645/files).  

If we want to then move this file into the cwl-v1.2 repository, we can then just generate a new PR that points to a new url

**Comment by mr-c** [comment]
Did you want to rewrite the commits, or shall I squash into one?

**Comment by alexiswl** [comment]
I can squash into one and force push if you'd like?

**Comment by alexiswl** [comment]
Done

**Pull request merged**